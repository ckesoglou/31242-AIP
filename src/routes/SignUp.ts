import { Request, Response, Router } from 'express';
import { BAD_REQUEST, CREATED, OK, UNPROCESSABLE_ENTITY } from 'http-status-codes';
import bcrypt from 'bcrypt';
import Joi, { ObjectSchema } from "joi";

import { createUser, getUser } from "@daos/Users";
import User from '@entities/User.ts';
import { generateNewAuthenticationTokens } from '@shared/Authenticate';

const router = Router();
//Salt Rounds determines how many times the password is hashed. 
//As bCrypt randomizes salts, this makes brute forcing and rainbow table attacks much more difficult.
const saltRounds = 10;

/**
 * POST: /signup
 */

interface ISignUpPOST {
  username: string;
  displayName: string;
  password: string;
}

const SignUpPOST: ObjectSchema<ISignUpPOST> = Joi.object({
  username: Joi.string().alphanum().min(2).max(16, "utf8"),
  password: Joi.string().min(8).max(72, "utf8")
    .ruleset.pattern(/(?=.*[A-Z])(?=.*[!@#$%^&*()_+\-=[\]{};':"\\|,.<>/?])/) // at least one uppercase and one symbol
    .rule({ message: "Password must have at least one uppercase character and symbol" }),
  displayName: Joi.string().trim().min(2).max(16, "utf8")
    .ruleset.pattern(/^[a-zA-Z0-9 ]*$/) // letters numbers and spaces
    .rule({ message: "Display name can only contain letters, numbers and spaces" }),
}).options({ presence: "required" });

router.post('/', async (req: Request, res: Response) => {
  const { error, value } = SignUpPOST.validate(req.body);

  if (error) {
    return res.status(BAD_REQUEST).json({
      errors: [error.message],
    });
  }

  const requestBody = value as ISignUpPOST;

  const foundUser = await getUser(requestBody.username);
  if (foundUser) {
    return res.status(UNPROCESSABLE_ENTITY).json({
      errors: ["Username already exists"],
    });
  }

  const passHash: string = await bcrypt.hash(requestBody.password, saltRounds)

  const user = await createUser(requestBody.username, requestBody.displayName, passHash);
  await generateNewAuthenticationTokens(user, req.headers.host ?? "Unknown", res);

  return res.status(CREATED).end();
});

export default router;