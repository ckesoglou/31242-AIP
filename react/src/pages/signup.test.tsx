import React from "react";
import { shallow, mount } from "enzyme";
import { MemoryRouter, Route } from "react-router-dom";
import SignUp from "./signup";

const testProps = {
  history: {} as any,
  location: {} as any,
  match: {} as any,
};

describe("SignUp", () => {
  it("should render correctly", () => {
    const component = shallow(
      <SignUp
        history={testProps.history}
        location={testProps.location}
        match={testProps.match}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it("navigates to next page for successful sign up", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/signup"]}>
        <Route path="/signup" component={SignUp} />
        <Route path="/home">
          <h1>Hello</h1>
        </Route>
      </MemoryRouter>
    );

    let signUpComponent = wrapper.find("SignUp");
    signUpComponent.setState({
      username: "",
      display_name: "",
      password: "",
      error: "",
      successfulSignUp: true,
    });

    // @ts-ignore
    expect(wrapper.find("Router").prop("history").location.pathname).toEqual(
      "/home"
    );
  });

  it("button should not be clickable on load", () => {
    const spy = jest.spyOn(SignUp.prototype, "handleSignUp");
    const wrapper = mount(
      <MemoryRouter>
        <SignUp
          history={testProps.history}
          location={testProps.location}
          match={testProps.match}
        />
      </MemoryRouter>
    );

    let button = wrapper.find("button");
    button.simulate("click");

    expect(spy).toHaveBeenCalledTimes(0);
    expect(button.html().includes('disabled=""')).toBe(true);
  });

  it("should handle sign up if button is clicked and information is filled", () => {
    const spy = jest.spyOn(SignUp.prototype, "handleSignUp");
    const wrapper = mount(
      <MemoryRouter>
        <SignUp
          history={testProps.history}
          location={testProps.location}
          match={testProps.match}
        />
      </MemoryRouter>
    );

    wrapper
      .find("input#password")
      .simulate("change", { target: { value: "This is a valid password!" } });
    wrapper
      .find("input#username")
      .simulate("change", { target: { value: "This is a valid username!" } });
    wrapper.find("input#display_name").simulate("change", {
      target: { value: "This is a valid display name!" },
    });

    let button = wrapper.find("button");
    button.simulate("click");

    expect(spy).toHaveBeenCalledTimes(1);
    expect(button.html().includes('disabled=""')).toBe(true);
  });
});
