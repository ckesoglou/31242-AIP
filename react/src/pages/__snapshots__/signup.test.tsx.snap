// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp should render correctly 1`] = `
<WithStyles(ForwardRef(Container))
  component="main"
  maxWidth="xs"
>
  <div
    className="paper"
  >
    <WithStyles(ForwardRef(Typography))
      component="h1"
      id="header"
      variant="h4"
    >
      Sign Up
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(FormControl))
      className="form"
    >
      <WithStyles(ForwardRef(TextField))
        fullWidth={true}
        id="username"
        label="User Name"
        margin="normal"
        name="username"
        onChange={[Function]}
        onFocus={[Function]}
        required={true}
        value=""
        variant="outlined"
      />
      <WithStyles(ForwardRef(TextField))
        fullWidth={true}
        id="display_name"
        label="Display Name"
        margin="normal"
        name="display_name"
        onChange={[Function]}
        onFocus={[Function]}
        required={true}
        value=""
        variant="outlined"
      />
      <WithStyles(ForwardRef(TextField))
        error={false}
        fullWidth={true}
        id="password"
        label="Password"
        margin="normal"
        name="password"
        onChange={[Function]}
        onFocus={[Function]}
        required={true}
        type="password"
        value=""
        variant="outlined"
      />
      <div
        id="passwordRequirements"
      >
        <ForwardRef(Grow)
          in={false}
          timeout={500}
        >
          <WithStyles(ForwardRef(Paper))
            id="passwordCriteria"
          >
            <p
              id="passwordCriteriaHeader"
            >
              Password Criteria
            </p>
            <p
              id="characterLength"
            >
              Needs to be at least 8 characters
            </p>
            <p
              id="uppercaseCharacter"
            >
              Needs to have one upper-case character
            </p>
            <p
              id="specialCharacter"
            >
              Needs to have one special character - e.g. !@#$%
            </p>
          </WithStyles(ForwardRef(Paper))>
        </ForwardRef(Grow)>
      </div>
      <WithStyles(ForwardRef(Button))
        color="primary"
        disabled={true}
        fullWidth={true}
        id="submit"
        onClick={[Function]}
        size="large"
        type="submit"
        variant="contained"
      >
        <WithStyles(ForwardRef(CircularProgress))
          color="inherit"
          id="loading"
          size={35}
        />
        <label
          id="signUpText"
        >
          Sign Up
        </label>
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Grid))
        container={true}
      >
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <WithStyles(ForwardRef(Link))
            href="/"
          >
            Sign in instead
          </WithStyles(ForwardRef(Link))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(FormControl))>
  </div>
</WithStyles(ForwardRef(Container))>
`;
